File: JavaScript/0701. Insert into a Binary Search Tree.js
// Given the root node of a binary search tree (BST) and a value to be inserted into the tree, insert the value into the BST. Return the root node of the BST after the insertion. It is guaranteed that the new value does not exist in the original BST.
// Note that there may exist multiple valid ways for the insertion, as long as the tree remains a BST after insertion. You can return any of them.
//
// For example,
// Given the tree:
//
//         4
//        / \
//       2   7
//      / \
//     1   3
//
// And the value to insert: 5
// You can return this binary search tree:
//
//          4
//        /   \
//       2     7
//      / \   /
//     1   3 5
//
// This tree is also valid:
//
//          5
//        /   \
//       2     7
//      / \
//     1   3
//          \
//           4

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */

/** 1) Recursion */
const insertIntoBST1 = (root, val) => {
  if (root == null) return new TreeNode(val);
  if (val < root.val) {
    root.left = insertIntoBST(root.left, val);
  } else {
    root.right = insertIntoBST(root.right, val);
  }
  return root;
};

/** 2) Iteration */
const insertIntoBST = (root, val) => {
  let node = root;
  while (node != null) {
    if (val < node.val) {
      // insert right now
      if (node.left == null) {
        node.left = new TreeNode(val);
        return root;
      } else {
        node = node.left;
      }
    } else {
      // insert right now
      if (node.right == null) {
        node.right = new TreeNode(val);
        return root;
      } else {
        node = node.right;
      }
    }
  }
  return new TreeNode(val);
};

File: Python/0701. Insert into a Binary Search Tree.py
# You are given the root node of a binary search tree (BST) and a value to insert into the tree. Return the root node of the BST after the insertion. It is guaranteed that the new value does not exist in the original BST.
# Notice that there may exist multiple valid ways for the insertion, as long as the tree remains a BST after insertion. You can return any of them.
#
# Example 1:
#
# Input: root = [4,2,7,1,3], val = 5
# Output: [4,2,7,1,3,5]
# Explanation: Another accepted tree is:
#
# Example 2:
#
# Input: root = [40,20,60,10,30,50,70], val = 25
# Output: [40,20,60,10,30,50,70,null,null,25]
# Example 3:
#
# Input: root = [4,2,7,1,3,null,null,null,null,null,null], val = 5
# Output: [4,2,7,1,3,5]
#
# Constraints:
#
# The number of nodes in the tree will be in the range [0, 10^4].
# -10^8 <= Node.val <= 10^8
# All the values Node.val are unique.
# -10^8 <= val <= 10^8
# It's guaranteed that val does not exist in the original BST.


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 1) Recursion
# Time O(H): O(N) worst case, O(logN) average case if the tree is well organized.
# Space O(H): O(N) worst case, O(logN) average case.
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return TreeNode(val)
        if root.val > val:
            root.left = self.insertIntoBST(root.left, val)
        else:
            root.right = self.insertIntoBST(root.right, val)
        return root


# 2) Iteration
# Time O(H): O(N) worst case, O(logN) average case if the tree is well organized.
# Space O(1)
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        node = root
        while node:
            if node.val > val:
                if not node.left:
                    node.left = TreeNode(val)
                    return root
                node = node.left
            else:
                if not node.right:
                    node.right = TreeNode(val)
                    return root
                node = node.right
        return TreeNode(val)
